pkg/local-gadget-manager/local-gadget-manager.go

```diff
@@ -0,0 +1,409 @@
+func (l *LocalGadgetManager) Unsubscribe(key interface{}) {
+	l.pubsub.Unsubscribe(key)
+}
+
+func (l *LocalGadgetManager) AddContainer(notif runcfanotify.ContainerEvent) {
+	pid := int(notif.ContainerPID)
+
+	cgroupPathV1, cgroupPathV2, err := containerutils.GetCgroupPaths(pid)
+	if err != nil {
+		return
+	}
+	cgroupPathV2WithMountpoint, _ := containerutils.CgroupPathV2AddMountpoint(cgroupPathV2)
+	cgroupId, _ := containerutils.GetCgroupID(cgroupPathV2WithMountpoint)
+	mntns, err := containerutils.GetMntNs(pid)
+
+	containerDefinition := pb.ContainerDefinition{
+		Id:         notif.ContainerID,
+		Pid:        notif.ContainerPID,
+		Namespace:  "local",
+		Podname:    notif.ContainerID,
+		CgroupPath: cgroupPathV2WithMountpoint,
+		CgroupId:   cgroupId,
+		Mntns:      mntns,
+		CgroupV1:   cgroupPathV1,
+		CgroupV2:   cgroupPathV2,
+	}
+
+	if l.dockercli != nil {
+		filter := filters.NewArgs()
+		filter.Add("id", notif.ContainerID)
+		containers, err := l.dockercli.ContainerList(context.Background(),
+			dockertypes.ContainerListOptions{
+				All:     true,
+				Filters: filter,
+			})
+		if err != nil {
+			fmt.Printf("Couldn't find docker container: %s", err)
+		}
+		if len(containers) == 1 {
+			if len(containers[0].Names) > 0 {
+				containerDefinition.Podname = strings.TrimPrefix(containers[0].Names[0], "/")
+			}
+		}
+	}
+
+	l.containers[notif.ContainerID] = containerDefinition
+
+	l.pubsub.Publish(pubsub.EVENT_TYPE_ADD_CONTAINER, containerDefinition)
+
+}
+
+func (l *LocalGadgetManager) RemoveContainer(notif runcfanotify.ContainerEvent) {
+	containerDefinition, ok := l.containers[notif.ContainerID]
+	if !ok {
+		return
+	}
+	l.pubsub.Publish(pubsub.EVENT_TYPE_REMOVE_CONTAINER, containerDefinition)
+	delete(l.containers, notif.ContainerID)
+}
+
+func (l *LocalGadgetManager) Dump() string {
+	out := "List of containers:\n"
+	for i, c := range l.containers {
+		out += fmt.Sprintf("%v -> %+v\n", i, c)
+	}
+	out += "List of tracers:\n"
+	for i, t := range l.tracers {
+		out += fmt.Sprintf("%v -> %q %q\n",
+			i,
+			t.gadget,
+			t.name)
+		out += fmt.Sprintf("    %+v\n", t.traceResource)
+		out += fmt.Sprintf("    %+v\n", t.traceResource.Spec.Filter)
+	}
+	return out
+}
+func NewManager() (*LocalGadgetManager, error) {
+	l := &LocalGadgetManager{
+		traceFactories: gadgetcollection.TraceFactories(),
+		tracers:        make(map[string]tracer),
+		containers:     make(map[string]pb.ContainerDefinition),
+		pubsub:         pubsub.NewGadgetPubSub(),
+	}
+
+	for _, factory := range l.traceFactories {
+		factory.Initialize(l, nil)
+	}
+
+	runcNotifier, err := runcfanotify.NewRuncNotifier(func(notif runcfanotify.ContainerEvent) {
+		switch notif.Type {
+		case runcfanotify.EVENT_TYPE_ADD_CONTAINER:
+			l.AddContainer(notif)
+		case runcfanotify.EVENT_TYPE_REMOVE_CONTAINER:
+			l.RemoveContainer(notif)
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+	l.runcNotifier = runcNotifier
+
+	l.dockercli, _ = client.NewClientWithOpts(client.FromEnv)
```

Comment:
If I understood correctly the option `[container-filter]`, maybe we could check the error here to notify that filtering will not work?

Reply:
If the Docker API does not work, the container-filter falls back on the Container ID, so filtering can still work when the user passes the Container ID...

But I agree it is not a good semantic and should be changed.

Reply:
Understood.

I am thinking that even though we adopt `--container-id` and `--container-name` notation, if the container engine initialisation fails (In this case `client.NewClientWithOpts`), it means that `--container-name` will not work. I think we should at least notify that or what about not initialising the container engine here but directly in `AddContainer`? In the end, it is only used there. Of course, this is valid only if such initialisation is not too costly and there really exist the possibility that it fails 😄.